readthedocs/invitations/views.py

```diff
@@ -0,0 +1,112 @@
+"""Invitation views."""
+from django.contrib.auth.mixins import UserPassesTestMixin
+from django.http import Http404, HttpResponseBadRequest, HttpResponseRedirect
+from django.shortcuts import get_object_or_404
+from django.urls import reverse
+from django.views.generic import DeleteView, DetailView
+
+from readthedocs.core.mixins import PrivateViewMixin
+from readthedocs.core.permissions import AdminPermission
+from readthedocs.invitations.models import Invitation
+from readthedocs.organizations.models import Organization, Team, TeamInvite
+
+
+class RevokeInvitation(PrivateViewMixin, UserPassesTestMixin, DeleteView):
+
+    model = Invitation
+    pk_url_kwarg = "invitation_pk"
+    http_method_names = ["post"]
+
+    def test_func(self):
+        invitation = self.get_object()
+        return invitation.can_revoke_invitation(self.request.user)
+
+    def get_success_url(self):
+        invitation = self.get_object()
+        return invitation.get_origin_url()
+
+
+class RedeemInvitation(DetailView):
+
+    """
+    Allow the user to accept or decline an invitation.
+
+    The invitation is deleted once it has been accepted or declined.
+    """
+
+    model = Invitation
+    slug_url_kwarg = "invitation_token"
+    slug_field = "token"
+    context_object_name = "invitation"
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        invitation = self.get_object()
+        from_user = invitation.from_user
+        context["requestor_name"] = from_user.get_full_name() or from_user.username
+        context["requestor_url"] = reverse(
+            "profiles_profile_detail", args=[from_user.username]
+        )
+        context["target_name"] = invitation.object_name
+        context["target_type"] = invitation.object._meta.verbose_name
+        return context
+
+    # pylint: disable=unused-argument
+    def post(self, request, *args, **kwargs):
+        invitation = self.get_object()
+        url = reverse("homepage")
+        if request.POST.get("accept"):
+            if invitation.expired:
+                return HttpResponseBadRequest("Invitation has expired.")
+
+            # If the invitation is attached to an email,
+            # and the current user isn't logged-in we
+            # redeem the invitation after the user has signed-up.
+            if not request.user.is_authenticated and invitation.to_email:
+                return self.redeem_at_sign_up(invitation)
+            invitation.redeem(request.user)
+            url = invitation.get_success_url()
+        invitation.delete()
```

Comment:
It would be good to log that the invitation was declined here.

Reply:
I think we should keep the database object and mark it revoked. It should not be possible to accept an invitation and then have it deleted afterwards.

Reply:
@stsewd do you want me to migrate this input to a follow-up issue? I think it fits well together with the few other Invitation model updates suggested.